### PR TITLE
Deduplicate functions after provider merge

### PR DESCRIPTION
**Pull request by carlosedp**
## Description

The use of containerd and CNI functions has been refactored to reuse
the same codebase.

Added all network functionality to own directory and package.

Rename containers handler to functions to clear-up functionality.

Persist data between reboots storing files in /var/lib/faasd

## Motivation and Context

- [ ] I have raised an issue to propose this change **this is required**

This continues the work from #25.

## How Has This Been Tested?

Started both faasd and provider, issued deploy, list, remove and scale commands. Also tested deployed functions.

Cli:

```
❯ faas-cli list
Function                      	Invocations    	Replicas
figlet1                       	0              	1
❯ curl -d Test http://127.0.0.1:8080/function/figlet1
 _____         _
|_   _|__  ___| |_
  | |/ _ \/ __| __|
  | |  __/\__ \ |_
  |_|\___||___/\__|

❯
❯ faas store deploy figlet --name=figlet2 --env write_timeout=1s
WARNING! Communication is not secure, please consider using HTTPS. Letsencrypt.org offers free SSL/TLS certificates.

Deployed. 200 OK.
URL: http://127.0.0.1:8080/function/figlet2

❯ curl -d Test http://127.0.0.1:8080/function/figlet2
 _____         _
|_   _|__  ___| |_
  | |/ _ \/ __| __|
  | |  __/\__ \ |_
  |_|\___||___/\__|

❯ faas-cli list
Function                      	Invocations    	Replicas
figlet1                       	1              	1
figlet2                       	0              	1

❯ faas-cli remove figlet2
Deleting: figlet2.
Removing old function.
❯ faas-cli list
Function                      	Invocations    	Replicas
figlet1                       	1              	1

```

On faasd:

```
Feb 03 14:04:25 debian10 systemd[1]: Started faasd.
-- Subject: A start job for unit faasd.service has finished successfully
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- A start job for unit faasd.service has finished successfully.
--
-- The job identifier is 8845.
Feb 03 14:04:25 debian10 faasd[12295]: 2020/02/03 14:04:25 File exists: "/var/lib/faasd/secrets/basic-auth-password"
Feb 03 14:04:25 debian10 faasd[12295]: 2020/02/03 14:04:25 File exists: "/var/lib/faasd/secrets/basic-auth-user"
Feb 03 14:04:25 debian10 faasd[12295]: 2020/02/03 14:04:25 Writing network config...
Feb 03 14:04:25 debian10 faasd[12295]: 2020/02/03 14:04:25 Supervisor created in: 996.891µs
Feb 03 14:04:25 debian10 faasd[12295]: Preparing: basic-auth-plugin with image: docker.io/openfaas/basic-auth-plugin:0.18.10
Feb 03 14:04:25 debian10 faasd[12295]: Prepare done for: docker.io/openfaas/basic-auth-plugin:0.18.10, 7171180 bytes
Feb 03 14:04:25 debian10 faasd[12295]: Preparing: nats with image: docker.io/library/nats-streaming:0.11.2
Feb 03 14:04:25 debian10 faasd[12295]: Prepare done for: docker.io/library/nats-streaming:0.11.2, 4647125 bytes
Feb 03 14:04:25 debian10 faasd[12295]: Preparing: prometheus with image: docker.io/prom/prometheus:v2.14.0
Feb 03 14:04:25 debian10 faasd[12295]: Prepare done for: docker.io/prom/prometheus:v2.14.0, 53527559 bytes
Feb 03 14:04:25 debian10 faasd[12295]: Preparing: gateway with image: docker.io/openfaas/gateway:0.18.8
Feb 03 14:04:25 debian10 faasd[12295]: Prepare done for: docker.io/openfaas/gateway:0.18.8, 11117867 bytes
Feb 03 14:04:25 debian10 faasd[12295]: Preparing: queue-worker with image: docker.io/openfaas/queue-worker:0.9.0
Feb 03 14:04:26 debian10 faasd[12295]: Prepare done for: docker.io/openfaas/queue-worker:0.9.0, 4293659 bytes
Feb 03 14:04:26 debian10 faasd[12295]: Reconciling: basic-auth-plugin
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 Created container basic-auth-plugin
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 [CNI] getting IP address for container: basic-auth-plugin PID: 12341
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 basic-auth-plugin has IP: 10.62.1.51
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 Task: basic-auth-plugin        Container: basic-auth-plugin
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 19:04:26 Listening on: 8080
Feb 03 14:04:26 debian10 faasd[12295]: Reconciling: nats
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 Created container nats
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 [CNI] getting IP address for container: nats PID: 12439
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 nats has IP: 10.62.1.52
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 Task: nats        Container: nats
Feb 03 14:04:26 debian10 faasd[12295]: Reconciling: prometheus
Feb 03 14:04:26 debian10 faasd[12295]: Status of prometheus is: stopped
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 Need to kill prometheus
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.750100 [INF] STREAM: Starting nats-streaming-server[faas-cluster] version 0.11.2
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.750141 [INF] STREAM: ServerID: xKzC406TJ2pSKda6bEQcau
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.750143 [INF] STREAM: Go version: go1.11.1
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.751034 [INF] Starting nats-server version 1.3.0
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.751048 [INF] Git commit [not set]
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.751172 [INF] Starting http monitor on 0.0.0.0:8222
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.751554 [INF] Listening for client connections on 0.0.0.0:4222
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.751610 [INF] Server is ready
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.778794 [INF] STREAM: Recovering the state...
Feb 03 14:04:26 debian10 faasd[12295]: [1] 2020/02/03 19:04:26.778805 [INF] STREAM: No recovered state
Feb 03 14:04:26 debian10 faasd[12295]: 2020/02/03 14:04:26 Created container prometheus
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029681 [INF] STREAM: Message store is MEMORY
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029708 [INF] STREAM: ---------- Store Limits ----------
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029712 [INF] STREAM: Channels:                  100 *
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029714 [INF] STREAM: --------- Channels Limits --------
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029715 [INF] STREAM:   Subscriptions:          1000 *
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029717 [INF] STREAM:   Messages     :       1000000 *
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029719 [INF] STREAM:   Bytes        :     976.56 MB *
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029720 [INF] STREAM:   Age          :     unlimited *
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029722 [INF] STREAM:   Inactivity   :     unlimited *
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.029723 [INF] STREAM: ----------------------------------
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 [CNI] getting IP address for container: prometheus PID: 12555
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 prometheus has IP: 10.62.1.53
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 Task: prometheus        Container: prometheus
Feb 03 14:04:27 debian10 faasd[12295]: Reconciling: gateway
Feb 03 14:04:27 debian10 faasd[12295]: Status of gateway is: running
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 Need to kill gateway
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.248Z caller=main.go:296 msg="no time or size retention was set so using the default time retention" duration=15d
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.248Z caller=main.go:332 msg="Starting Prometheus" version="(version=2.14.0, branch=HEAD, revision=edeb7a44cbf745f1d8be4ea6f215e79e651bfe19)"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.248Z caller=main.go:333 build_context="(go=go1.13.4, user=root@df2327081015, date=20191111-14:27:12)"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.248Z caller=main.go:334 host_details="(Linux 4.19.0-6-amd64 #1 SMP Debian 4.19.67-2+deb10u2 (2019-11-11) x86_64 debian10 (none))"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.249Z caller=main.go:335 fd_limits="(soft=1024, hard=1024)"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.249Z caller=main.go:336 vm_limits="(soft=unlimited, hard=unlimited)"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.251Z caller=main.go:657 msg="Starting TSDB ..."
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.251Z caller=web.go:496 component=web msg="Start listening for connections" address=0.0.0.0:9090
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.254Z caller=head.go:535 component=tsdb msg="replaying WAL, this may take awhile"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.254Z caller=head.go:583 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.255Z caller=main.go:672 fs_type=794c7630
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.255Z caller=main.go:673 msg="TSDB started"
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.255Z caller=main.go:743 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.264Z caller=main.go:771 msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus.yml
Feb 03 14:04:27 debian10 faasd[12295]: level=info ts=2020-02-03T19:04:27.264Z caller=main.go:626 msg="Server is ready to receive web requests."
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 Created container gateway
Feb 03 14:04:27 debian10 faasd[12295]: [1] 2020/02/03 19:04:27.396039 [INF] STREAM: Channel "faas-request" has been created
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 [CNI] getting IP address for container: gateway PID: 12680
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 gateway has IP: 10.62.1.54
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 Task: gateway        Container: gateway
Feb 03 14:04:27 debian10 faasd[12295]: Reconciling: queue-worker
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 19:04:27 HTTP Read Timeout: 1m0s
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 19:04:27 HTTP Write Timeout: 1m0s
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 19:04:27 Binding to external function provider: http://faas-containerd:8081/
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 19:04:27 Async enabled: Using NATS Streaming.
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 19:04:27 Opening connection to nats://nats:4222
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 19:04:27 Connect: nats://nats:4222
Feb 03 14:04:27 debian10 faasd[12295]: Status of queue-worker is: running
Feb 03 14:04:27 debian10 faasd[12295]: 2020/02/03 14:04:27 Need to kill queue-worker
Feb 03 14:04:28 debian10 faasd[12295]: 2020/02/03 14:04:28 Created container queue-worker
Feb 03 14:04:28 debian10 faasd[12295]: 2020/02/03 14:04:28 [CNI] getting IP address for container: queue-worker PID: 12793
Feb 03 14:04:28 debian10 faasd[12295]: 2020/02/03 14:04:28 queue-worker has IP: 10.62.1.55
Feb 03 14:04:28 debian10 faasd[12295]: 2020/02/03 14:04:28 Task: queue-worker        Container: queue-worker
Feb 03 14:04:28 debian10 faasd[12295]: 2020/02/03 14:04:28 Supervisor init done in: 2.365384679s
Feb 03 14:04:28 debian10 faasd[12295]: Loading basic authentication credentials
Feb 03 14:04:28 debian10 faasd[12295]: 2020/02/03 14:04:28 faasd: waiting for SIGTERM or SIGINT
Feb 03 14:04:28 debian10 faasd[12295]: Connect: nats://nats:4222
Feb 03 14:04:28 debian10 faasd[12295]: Subscribing to: faas-request at nats://nats:4222
Feb 03 14:04:28 debian10 faasd[12295]: Wait for  5m5s
Feb 03 14:04:28 debian10 faasd[12295]: Listening on [faas-request], clientID=[faas-worker-debian10], qgroup=[faas] durable=[]
Feb 03 14:04:31 debian10 faasd[12295]: 2020/02/03 14:04:31 [up] Sending 10.62.1.54 to proxy
Feb 03 14:04:31 debian10 faasd[12295]: 2020/02/03 14:04:31 Starting faasd proxy on 8080
Feb 03 14:04:31 debian10 faasd[12295]: Gateway: 10.62.1.54:8080
Feb 03 14:04:31 debian10 faasd[12295]: 2020/02/03 14:04:31 [proxy] Wait for done
Feb 03 14:04:31 debian10 faasd[12295]: 2020/02/03 14:04:31 [proxy] Begin listen on 8080
Feb 03 14:07:55 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/system/functions
Feb 03 14:07:55 debian10 faasd[12295]: 2020/02/03 19:07:55 Validated request 200.

Feb 03 14:08:00 debian10 faasd[12295]: 2020/02/03 19:08:00 Get http://faas-containerd:8081/system/functions: dial tcp: i/o timeout
Feb 03 14:08:00 debian10 faasd[12295]: 2020/02/03 19:08:00 Forwarded [GET] to /system/functions - [200] - 5.042371 seconds
Feb 03 14:08:05 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/function/figlet1
Feb 03 14:08:05 debian10 faasd[12295]: 2020/02/03 19:08:05 GetReplicas [figlet1.] took: 0.011471s
Feb 03 14:08:05 debian10 faasd[12295]: 2020/02/03 19:08:05 GetReplicas [figlet1.] took: 0.011491s
Feb 03 14:08:05 debian10 faasd[12295]: 2020/02/03 19:08:05 Forwarded [POST] to /function/figlet1 - [200] - 0.007468 seconds
Feb 03 14:08:14 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/system/functions
Feb 03 14:08:14 debian10 faasd[12295]: 2020/02/03 19:08:14 Validated request 200.
Feb 03 14:08:14 debian10 faasd[12295]: 2020/02/03 19:08:14 Forwarded [PUT] to /system/functions - [404] - 0.001887 seconds
Feb 03 14:08:14 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/system/functions
Feb 03 14:08:14 debian10 faasd[12295]: 2020/02/03 19:08:14 Validated request 200.
Feb 03 14:08:15 debian10 faasd[12295]: 2020/02/03 19:08:15 Forwarded [POST] to /system/functions - [200] - 0.400285 seconds
Feb 03 14:08:20 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/function/figlet2
Feb 03 14:08:20 debian10 faasd[12295]: 2020/02/03 19:08:20 GetReplicas [figlet2.] took: 0.012061s
Feb 03 14:08:20 debian10 faasd[12295]: 2020/02/03 19:08:20 GetReplicas [figlet2.] took: 0.012079s
Feb 03 14:08:20 debian10 faasd[12295]: 2020/02/03 19:08:20 Forwarded [POST] to /function/figlet2 - [200] - 0.010504 seconds
Feb 03 14:08:22 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/system/functions
Feb 03 14:08:22 debian10 faasd[12295]: 2020/02/03 19:08:22 Validated request 200.
Feb 03 14:08:22 debian10 faasd[12295]: 2020/02/03 19:08:22 Forwarded [GET] to /system/functions - [200] - 0.009720 seconds
Feb 03 14:08:30 debian10 faasd[12295]: 2020/02/03 19:08:30 Get http://faas-containerd:8081/system/functions: dial tcp: i/o timeout
Feb 03 14:09:20 debian10 faasd[12295]: 2020/02/03 19:09:20 Get http://faas-containerd:8081/system/functions: dial tcp: i/o timeout
Feb 03 14:11:47 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/system/functions
Feb 03 14:11:47 debian10 faasd[12295]: 2020/02/03 19:11:47 Validated request 200.
Feb 03 14:11:49 debian10 faasd[12295]: 2020/02/03 19:11:49 Forwarded [DELETE] to /system/functions - [200] - 2.418039 seconds
Feb 03 14:11:50 debian10 faasd[12295]: [faasd] proxy: http://10.62.1.54:8080/system/functions
Feb 03 14:11:50 debian10 faasd[12295]: 2020/02/03 19:11:50 Validated request 200.
Feb 03 14:11:50 debian10 faasd[12295]: 2020/02/03 19:11:50 Forwarded [GET] to /system/functions - [200] - 0.007062 seconds
```

On faasd-provider:

```
Feb 03 14:04:25 debian10 systemd[1]: Started faasd-provider.
-- Subject: A start job for unit faasd-provider.service has finished successfully
-- Defined-By: systemd
-- Support: https://www.debian.org/support
--
-- A start job for unit faasd-provider.service has finished successfully.
--
-- The job identifier is 8920.
Feb 03 14:04:25 debian10 faasd[12303]: 2020/02/03 14:04:25 faas-containerd starting..        Service Timeout: 1m0s
Feb 03 14:04:25 debian10 faasd[12303]: 2020/02/03 14:04:25 Writing network config...
Feb 03 14:04:25 debian10 faasd[12303]: 2020/02/03 14:04:25 Listening on TCP port: 8081
Feb 03 14:08:05 debian10 faasd[12303]: 2020/02/03 14:08:05 Resolve: "figlet1"
Feb 03 14:08:05 debian10 faasd[12303]: 2020/02/03 14:08:05 figlet1 took 0.002371 seconds
Feb 03 14:08:14 debian10 faasd[12303]: 2020/02/03 14:08:14 [Update] request: {"service":"figlet2","image":"functions/figlet:0.13.0","network":"","envProcess":"figlet","envVars":{"write_timeout":"1s"},"constraints":[],"secrets":[],"labels":{},"annotations":{},"limits":null,"requests":null,"readOnlyRootFilesystem":false}
Feb 03 14:08:14 debian10 faasd[12303]: 2020/02/03 14:08:14 [Update] service figlet2 not found
Feb 03 14:08:14 debian10 faasd[12303]: 2020/02/03 14:08:14 [Deploy] request: {"service":"figlet2","image":"functions/figlet:0.13.0","network":"","envProcess":"figlet","envVars":{"write_timeout":"1s"},"constraints":[],"secrets":[],"labels":{},"annotations":{},"limits":null,"requests":null,"readOnlyRootFilesystem":false}
Feb 03 14:08:14 debian10 faasd[12303]: 2020/02/03 14:08:14 Deploy docker.io/functions/figlet:0.13.0 size: 5658006
Feb 03 14:08:15 debian10 faasd[12303]: 2020/02/03 14:08:15 Container ID: figlet2        Task ID figlet2:        Task PID: 12975
Feb 03 14:08:15 debian10 faasd[12303]: 2020/02/03 14:08:15 figlet2 has IP: 10.62.1.56.
Feb 03 14:08:15 debian10 faasd[12303]: 2020/02/03 19:08:15 Version: 0.13.0        SHA: fa93655d90d1518b04e7cfca7d7548d7d133a34e
Feb 03 14:08:15 debian10 faasd[12303]: 2020/02/03 19:08:15 Read/write timeout: 5s, 1s. Port: 8080
Feb 03 14:08:15 debian10 faasd[12303]: 2020/02/03 19:08:15 Writing lock-file to: /tmp/.lock
Feb 03 14:08:15 debian10 faasd[12303]: 2020/02/03 19:08:15 Metrics server. Port: 8081
Feb 03 14:08:20 debian10 faasd[12303]: 2020/02/03 14:08:20 Resolve: "figlet2"
Feb 03 14:08:20 debian10 faasd[12303]: 2020/02/03 19:08:20 Forking fprocess.
Feb 03 14:08:20 debian10 faasd[12303]: 2020/02/03 19:08:20 Wrote 120 Bytes - Duration: 0.002582 seconds
Feb 03 14:08:20 debian10 faasd[12303]: 2020/02/03 14:08:20 figlet2 took 0.004401 seconds
Feb 03 14:11:47 debian10 faasd[12303]: 2020/02/03 14:11:47 [Delete] request: {"functionName":"figlet2"}
Feb 03 14:11:47 debian10 faasd[12303]: 2020/02/03 14:11:47 [Delete] removing CNI network for: figlet2
Feb 03 14:11:47 debian10 faasd[12303]: 2020/02/03 14:11:47 [Delete] removed: figlet2 from namespace: /proc/12975/ns/net, ID: figlet2-12975
Feb 03 14:11:47 debian10 faasd[12303]: Status of figlet2 is: running
Feb 03 14:11:47 debian10 faasd[12303]: 2020/02/03 14:11:47 Need to kill figlet2
Feb 03 14:11:47 debian10 faasd[12303]: 2020/02/03 19:11:47 SIGTERM received.. shutting down server in 1s
Feb 03 14:11:47 debian10 faasd[12303]: 2020/02/03 19:11:47 Removing lock-file : /tmp/.lock
Feb 03 14:11:48 debian10 faasd[12303]: 2020/02/03 19:11:48 No new connections allowed. Exiting in: 1s
Feb 03 14:11:49 debian10 faasd[12303]: 2020/02/03 14:11:49 [Delete] deleted figlet2
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Commits:

- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] My commit message has a body and describe how this was tested and why it is required.
- [x] I have signed-off my commits with `git commit -s` for the Developer Certificate of Origin (DCO)

Code:

- [x] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.

Docs:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/cninetwork/cni_network.go [hunk 1]
```diff
@@ -165,6 +165,14 @@ func GetIPfromPID(pid int) (*net.IP, error) {
 
 }
 
+// CNIGateway returns the gateway for default subnet
+func CNIGateway() string {
+	ip, _, _ := net.ParseCIDR(defaultSubnet)
+	ip = ip.To4()
+	ip[3] = 1
```
Comment (alexellis) [inline pkg/cninetwork/cni_network.go]: Please use range checking. Return an error, don't swallow it.

File: pkg/provider/handlers/deploy.go [hunk 2]
```diff
@@ -83,17 +83,9 @@ func deploy(ctx context.Context, req types.FunctionDeployment, client *container
...
-		containerd.WithSnapshotter(snapshotter),
-		containerd.WithNewSnapshot(req.Service+"-snapshot", image),
-		containerd.WithNewSpec(oci.WithImageConfig(image),
-			oci.WithCapabilities([]string{"CAP_NET_RAW"}),
-			oci.WithMounts(mounts),
-			oci.WithEnv(envs)),
-	)
+	caps := []string{"CAP_NET_RAW"}
+
+	container, err := service.NewContainer(ctx, *client, name, image, snapshotter, caps, mounts, nil, envs)
```
Comment (alexellis) [inline pkg/provider/handlers/deploy.go]: I preferred the old version of this code, what's the reason for the helper method? What does the helper add ?

File: pkg/supervisor.go [hunk 3]
```diff
@@ -65,10 +47,15 @@ func (s *Supervisor) Close() {
...
 
 	for _, svc := range svcs {
-		err := service.Remove(ctx, s.client, svc.Name)
+		err := cninetwork.DeleteCNINetwork(ctx, s.cni, s.client, svc.Name)
+		if err != nil {
+			log.Printf("[Delete] error removing CNI network for %s, %s\n", svc.Name, err)
+		}
+
+		err = service.Remove(ctx, s.client, svc.Name)
 		if err != nil {
```
Comment (alexellis) [inline pkg/supervisor.go]: Why do we return here but not for the network error?

File: pkg/supervisor.go [hunk 4]
```diff
@@ -173,33 +146,18 @@ func (s *Supervisor) Start(svcs []Service) error {
...
-			return errors.Wrapf(err, "failed to setup network for namespace %q: %v", id, err)
+			return err
 		}
-
-		// Get the IP of the default interface.
-		// defaultInterface := gocni.DefaultPrefix + "0"
-		// ip := &result.Interfaces[defaultInterface].IPConfigs[0].IP
-		ip := getIP(newContainer.ID(), task.Pid())
-		log.Printf("%s has IP: %s\n", newContainer.ID(), ip)
+		log.Printf("%s has IP: %s.\n", newContainer.ID(), ip.String())
```
Comment (alexellis) [inline pkg/supervisor.go]: Remove the period in the log line

File: pkg/provider/handlers/functions.go [hunk 5]
```diff
@@ -61,7 +62,7 @@ func GetFunction(client *containerd.Client, name string) (Function, error) {
 				replicas = 1
 				f.pid = task.Pid()
 				// Get container IP address
-				ip, _ := GetIPfromPID(int(task.Pid()))
+				ip, _ := cninetwork.GetIPfromPID(int(task.Pid()))
```
Comment (alexellis) [inline pkg/provider/handlers/functions.go]: We can't swallow errors like this, so since you're working on this line. Please can you handle it?

**Comment by alexellis** [comment]
Given faas-containerd doesn't exist in this repo can you rename the title, it is a bit confusing as you saw earlier.

File: pkg/provider/handlers/deploy.go (hunk 2 above)
Comment (carlosedp) [reply pkg/provider/handlers/deploy.go]: Yes, I agree it's confusing.
Reverting back to call the NewContainer directly instead of the helper.

File: pkg/supervisor.go (hunk 3 above)
Comment (carlosedp) [reply pkg/supervisor.go]: Fixed.

File: pkg/supervisor.go (hunk 4 above)
Comment (carlosedp) [reply pkg/supervisor.go]: Done.

File: pkg/cninetwork/cni_network.go (hunk 1 above)
Comment (carlosedp) [reply pkg/cninetwork/cni_network.go]: Fixed.

File: pkg/provider/handlers/functions.go (hunk 5 above)
Comment (carlosedp) [reply pkg/provider/handlers/functions.go]: Done!

**Comment by alexellis** [comment]
Hi @carlosedp this needs a rebase, not sure if you were aware of that. Hope you can get to it soon. Let me know if the PR’s still needed too and I can keep it in my working memory.

Alex

**Comment by alexellis** [comment]
Thanks for the rebase. Could you re-test and update the console output please?

**Comment by carlosedp** [comment]
Evidences updated in the PR body. Pushed new changes and removed netlink/weave dependencies in favor of using CNI result files stored on the node.

File: pkg/cninetwork/cni_network.go (hunk 1 above)
Comment (alexellis) [reply pkg/cninetwork/cni_network.go]: I can't see any range-checking here at all. We may not have an index of [3]

File: pkg/provider/handlers/functions.go [hunk 6]
```diff
@@ -62,11 +63,10 @@ func GetFunction(client *containerd.Client, name string) (Function, error) {
 				f.pid = task.Pid()
 
 				// Get container IP address
-				ip, getIPErr := GetIPfromPID(int(task.Pid()))
```
Comment (alexellis) [inline pkg/provider/handlers/functions.go]: Why this change?

**Comment by alexellis** [comment]
New release for testing: https://github.com/alexellis/faasd/releases/tag/0.7.2

File: pkg/provider/handlers/functions.go (hunk 6 above)
Comment (carlosedp) [reply pkg/provider/handlers/functions.go]: To have all errors as `err`. I think it's easier visually.
